File: MatlabOctaveCodeSnippets/snip_frequency_not_bin_cent_cos.m
N=1024; n=(0:N-1)'; %number of samples and column vector
dw=2*pi/N; %FFT bin width
k1=115.3; k2=500.8; w1=k1*dw; w2=k2*dw;%define frequencies
x=10*cos(w1*n) + 1*cos(w2*n); %generate sum of two cosines
Sms = abs(fft(x)/N).^2; %MS spectrum |DTFS{x}|^2
[S,w] = periodogram(x); %Periodogram, frequency w=0 to pi
Power = sum(Sms) %estimate the power in Watts = A^2/2
k=0:N-1; %frequency index
subplot(211), h=plot(k,10*log10(Sms)); %MS spectrum (dB)
subplot(212), h2=plot(w,10*log10(S)); %Periodogram (dB)

File: PythonCodeSnippets/snip_frequency_not_bin_cent_cos.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import periodogram

N = 1024
n = np.arange(0, N)
dw = 2 * np.pi / N
k1, k2 = 115.3, 500.8
w1, w2 = k1 * dw, k2 * dw

x = 10 * np.cos(w1 * n) + 1 * np.cos(w2 * n)

Sms = np.abs(np.fft.fft(x) / N) ** 2
w = np.linspace(0, np.pi, len(Sms))

frequencies, S = periodogram(x)

Power = np.sum(Sms)

k = np.arange(0, N)

plt.subplot(211)
h = plt.plot(k, 10 * np.log10(Sms))
plt.title("MS Spectrum (dB)")

plt.subplot(212)
h2 = plt.plot(frequencies, 10 * np.log10(S))
plt.title("Periodogram (dB)")

plt.show()
